awx/ui_next/src/util/workflow.jsx

```diff
@@ -0,0 +1,204 @@
+    },
+  ];
+}
+
+export function getLinkOverlayPoints(d, nodePositions) {
+  const sourceX =
+    nodePositions[d.source.id].x + nodePositions[d.source.id].width + 1;
+  let sourceY =
+    normalizeY(nodePositions, nodePositions[d.source.id].y) +
+    nodePositions[d.source.id].height / 2;
+  const targetX = nodePositions[d.target.id].x - 1;
+  const targetY =
+    normalizeY(nodePositions, nodePositions[d.target.id].y) +
+    nodePositions[d.target.id].height / 2;
+
+  // There's something off with the math on the root node...
+  if (d.source.id === 1) {
+    sourceY += 10;
+  }
+  const slope = (targetY - sourceY) / (targetX - sourceX);
+  const yIntercept = targetY - slope * targetX;
+  const orthogonalDistance = 8;
+
+  const pt1 = [
+    targetX,
+    slope * targetX +
+      yIntercept +
+      orthogonalDistance * Math.sqrt(1 + slope * slope),
+  ].join(',');
+  const pt2 = [
+    sourceX,
+    slope * sourceX +
+      yIntercept +
+      orthogonalDistance * Math.sqrt(1 + slope * slope),
+  ].join(',');
+  const pt3 = [
+    sourceX,
+    slope * sourceX +
+      yIntercept -
+      orthogonalDistance * Math.sqrt(1 + slope * slope),
+  ].join(',');
+  const pt4 = [
+    targetX,
+    slope * targetX +
+      yIntercept -
+      orthogonalDistance * Math.sqrt(1 + slope * slope),
+  ].join(',');
+
+  return [pt1, pt2, pt3, pt4].join(' ');
+}
+
+export function layoutGraph(nodes, links) {
+  const g = new dagre.graphlib.Graph();
+  g.setGraph({ rankdir: 'LR', nodesep: 30, ranksep: 120 });
+
+  // This is needed for Dagre
+  g.setDefaultEdgeLabel(() => {
+    return {};
+  });
+
+  nodes.forEach(node => {
+    if (node.id === 1) {
+      g.setNode(node.id, {
+        label: '',
+        width: constants.rootW,
+        height: constants.rootH,
+      });
+    } else {
+      g.setNode(node.id, {
+        label: '',
+        width: constants.nodeW,
+        height: constants.nodeH,
+      });
+    }
+  });
+
+  links.forEach(link => {
+    g.setEdge(link.source.id, link.target.id);
+  });
+
+  dagre.layout(g);
+
+  return g;
+}
+
+export function getZoomTranslate(svgRef, newScale) {
+  const svgElement = document.getElementById('workflow-svg');
+  const svgBoundingClientRect = svgElement.getBoundingClientRect();
+  const current = d3.zoomTransform(d3.select(svgRef).node());
+  const origScale = current.k;
+  const unscaledOffsetX =
+    (current.x +
+      (svgBoundingClientRect.width * origScale - svgBoundingClientRect.width) /
+        2) /
+    origScale;
+  const unscaledOffsetY =
+    (current.y +
+      (svgBoundingClientRect.height * origScale -
+        svgBoundingClientRect.height) /
+        2) /
+    origScale;
+  const translateX =
+    unscaledOffsetX * newScale -
+    (newScale * svgBoundingClientRect.width - svgBoundingClientRect.width) / 2;
+  const translateY =
+    unscaledOffsetY * newScale -
+    (newScale * svgBoundingClientRect.height - svgBoundingClientRect.height) /
+      2;
+  return [translateX, translateY];
+}
```

Comment:
I get why we are doing this (share between the output and edit graphs), but it feels a bit weird in utils.

For the future it would be nice if all the graph-related stuff lived in components/workflows and that's the bulk of all the code and then the stuff in screens between Visualizer and Output were ver short files that maybe instatiate that combine the api endpoints, call the workflow graph components and maybe instatiate a prop that makes the graph act as edit/view.

Reply:
I moved this into components/Workflow